datafusion/physical-plan/src/joins/stream_join_utils.rs

```diff
@@ -740,20 +617,338 @@ pub fn record_visited_indices<T: ArrowPrimitiveType>(
+    RightExhausted,
+
+    /// State representing that the left side of the join has been fully processed.
+    LeftExhausted,
+
+    /// Represents a state where both sides of the join are exhausted.
+    ///
+    /// The `final_result` field indicates whether the join operation has
+    /// produced a final result or not.
+    BothExhausted { final_result: bool },
+}
+
+/// Represents the asynchronous trait for an eager join stream.
+/// This trait defines the core methods for handling asynchronous join operations
+/// between two streams (left and right).
+#[async_trait]
+pub trait EagerJoinStream {
+    /// Implements the main polling logic for the join stream.
+    ///
+    /// This method continuously checks the state of the join stream and
+    /// acts accordingly by delegating the handling to appropriate sub-methods
+    /// depending on the current state.
+    ///
+    /// # Arguments
+    ///
+    /// * `cx` - A context that facilitates cooperative non-blocking execution within a task.
+    ///
+    /// # Returns
+    ///
+    /// * `Poll<Option<Result<RecordBatch>>>` - A polled result, either a `RecordBatch` or None.
+    fn poll_next_impl(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<RecordBatch>>>
+    where
+        Self: Send,
+    {
+        loop {
+            return match self.state() {
+                EagerJoinStreamState::PullRight => {
+                    handle_async_state!(self.fetch_next_from_right_stream(), cx)
+                }
+                EagerJoinStreamState::PullLeft => {
+                    handle_async_state!(self.fetch_next_from_left_stream(), cx)
+                }
+                EagerJoinStreamState::RightExhausted => {
+                    handle_async_state!(self.handle_right_stream_end(), cx)
+                }
+                EagerJoinStreamState::LeftExhausted => {
+                    handle_async_state!(self.handle_left_stream_end(), cx)
+                }
+                EagerJoinStreamState::BothExhausted {
+                    final_result: false,
+                } => {
+                    handle_state!(self.prepare_for_final_results_after_exhaustion())
+                }
+                EagerJoinStreamState::BothExhausted { final_result: true } => {
+                    Poll::Ready(None)
+                }
+            };
+        }
+    }
+    /// Asynchronously pulls the next batch from the right (probe) stream.
+    ///
+    /// This default implementation checks for the next value in the right stream.
+    /// If a batch is found, the state is switched to `PullLeft`, and the batch handling
+    /// is delegated to `process_batch_from_right`. If the stream ends, the state is set to `RightExhausted`.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<StreamJoinStateResult<Option<RecordBatch>>>` - The state result after pulling the batch.
+    async fn fetch_next_from_right_stream(
```

Comment:
I am somewhat confused by this default implementation as it implies that join will always "ping pong" back and forth between fetching left and right inputs, while in realty I think the details of how the stream is implemented and how the join keys are distributed across batches could require fetching multiple batches from one (or both) inputs before progress can be made. 

I am thinking of  a join on `a = b` where all the rows in the batch have the same join key, for example:

Batch 1
| `a` |
|--------|
| 100 |
| 100 | 

Batch 2
| `a` |
|--------|
| 100 |
| 100 | 

Batch 3
| `a` |
|--------|
| 100 |
| 200 | 

Wouldn't the symmetric hash join have to read all three batches to find the next join key (`200`) before reading a batch from the other input / producing output?

I didn't see how the symmetric hash handles this case, so I must be missing something

Reply:
Both this and previous implementations buffer rows on both sides, adhering to deletion criteria set by the interval library; this aspect remains unchanged.

The core proposal focuses on retaining control over the SendableRecordBatch streams instead of merging them with futures::select. Additionally, there are plans to develop more efficient yielding strategies in the future. The suggested alternation strategy is expected to advance into sophisticated load-balancing techniques. This characteristic is fundamental to the realization of these advanced features.

Reply:
> Both this and previous implementations buffer rows on both sides, adhering to deletion criteria set by the interval library; this aspect remains unchanged.

I agree this PR doesn't seem to change any behavior. 

> Additionally, there are plans to develop more efficient yielding strategies in the future. The suggested alternation strategy is expected to advance into sophisticated load-balancing techniques.

Are these plans described anywhere?

Reply:
> Are these plans described anywhere?

Not yet, but hopefully soon. We will continue publishing blog posts about this stuff in Datafusion and will talk about future goodies there :)